### DISCUSSION TITLE
how to infer store type?

### DISCUSSION THREAD

**Discussion by vorant94** [pmndrs/zustand#2928] [Q&A] [answered]
with current generation of ts-friendly libs there is usually a way to infer type definitions from the lib data structures

for example with zod we can do something like this

```ts
export const categorySchema = z.object({
	id: z.number(),
	name: z.string().min(1),
	color: z.string(),
});
export type CategoryModel = z.infer<typeof categorySchema>;
```

with drizzle orm we can do something like this

```ts
export const users = sqliteTable("users", {
	id: text().primaryKey(),
	createdAt: text(),
	updatedAt: text().notNull(),
});
export type User = typeof users.$inferSelect;
```

is there a way to do the same with zustand? meaning instead of defining state interface and using it as a generic for `create` function,  call `create` function and infer the state type out of what was returned by it?

**Comment by dai-shi** [comment, marked as answer]
There would be several ways, but I'd do this:
```ts
const useMyStore = create(combine({
  count: 0,
}, (set) => ({
  inc: (by = 1) => set((state) => ({ count: state.count + by }));
}));

type MyState = ReturnType<(typeof useMyStore)['getState']>;
```

**Reply by vorant94** [reply]
thanks, that is exactly what i looked!

WDYT about adding some built-in syntax sugar for it like mentioned above libs? something like this maybe

```ts
type MyState = typeof useMyStore.$infer;
```

or

```ts
import type { InferState } from "zustand";

type MyState = InferState<typeof useMyStore>;
```

if its good, i can prepare a pr

**Reply by dai-shi** [reply]
Actually, we already have `ExtractState` internally.
If you want to export it, feel free to send a PR, but it requires to add documentation too.
(If we were to export it, we would like to move the definition from react.ts to vanilla.ts, as it's not react specific.)

**Reply by vorant94** [reply]
nice! will do in upcoming days

**Reply by dai-shi** [reply]
#2935
